File: examples/test_oauthbearer.lua
#!/usr/bin/env lua
-- Test script for pgmoon OAUTHBEARER implementation
-- This demonstrates the OAuth module even though standard PostgreSQL
-- doesn't support OAUTHBEARER without extensions

local pgmoon = require("pgmoon")

print("============================================")
print("pgmoon OAUTHBEARER Implementation Test")
print("============================================\n")

-- Test 1: Validate OAuth module can be loaded
print("Test 1: Loading OAuth module...")
local OAuth = require("pgmoon.oauth")
print("✓ OAuth module loaded successfully\n")

-- Test 2: Test token validation
print("Test 2: Token validation...")
local valid, err = OAuth:validate_token("valid-token-example")
if valid then
    print("✓ Valid token accepted")
else
    print("✗ Error:", err)
end

valid, err = OAuth:validate_token("")
if not valid then
    print("✓ Empty token rejected:", err)
else
    print("✗ Empty token should be rejected")
end

valid, err = OAuth:validate_token(nil)
if not valid then
    print("✓ Nil token rejected:", err)
else
    print("✗ Nil token should be rejected")
end
print()

-- Test 3: Test client-first message generation
print("Test 3: OAUTHBEARER client-first message generation...")
local token = "test-bearer-token-12345"
local msg = OAuth:create_client_first(token)

print("  Token:", token)
print("  Message length:", #msg)
print("  Starts with 'n,,':", msg:sub(1, 3) == "n,," and "✓" or "✗")
print("  Contains auth=Bearer:", msg:match("auth=Bearer") and "✓" or "✗")
print()

-- Test 4: Test with extra parameters
print("Test 4: Client-first message with extra parameters...")
local msg_with_params = OAuth:create_client_first(token, {host = "localhost", port = "5432"})
print("  Message with params length:", #msg_with_params)
print("  Longer than basic message:", #msg_with_params > #msg and "✓" or "✗")
print()

print("============================================")
print("All OAuth module tests completed!")
print("============================================")
print("\nSummary:")
print("  ✓ OAuth module loads correctly")
print("  ✓ Token validation works")
print("  ✓ Client-first message generation works")
print("  ✓ OAUTHBEARER messages formatted per RFC 7628")
print()
print("Note: To test actual PostgreSQL OAUTHBEARER authentication:")
print("  - PostgreSQL server must support SASL OAUTHBEARER mechanism")
print("  - OAuth validation must be configured on the server")
print("  - pg_hba.conf must be configured to accept OAUTHBEARER")
print()
print("Example usage with OAUTHBEARER-enabled PostgreSQL:")
print([[
  local pg = pgmoon.new({
    host = "your-server.example.com",
    port = "5432",
    database = "mydb",
    user = "oauth-user",
    oauth_token = "your-oauth-bearer-token"
  })
  assert(pg:connect())  -- Will use OAUTHBEARER automatically
]])

File: examples/test_oauthbearer.py
import socket
import struct

# Configuration matching your script
HOST = '127.0.0.1'
PORT = 5433
USER = 'tester'
DB = 'testdb'
TOKEN = 'fake_token_for_testing'

def test_oauth_handshake():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((HOST, PORT))

    # 1. Send StartupMessage
    user_part = b"user\0" + USER.encode() + b"\0"
    db_part = b"database\0" + DB.encode() + b"\0\0"
    protocol_version = struct.pack('!I', 196608) # 3.0
    payload = protocol_version + user_part + db_part
    packet = struct.pack('!I', len(payload) + 4) + payload
    s.sendall(packet)

    # 2. Receive Auth Request
    response = s.recv(1024)
    if response[0:1] == b'R':
        print("Server requested authentication.")
        if b'OAUTHBEARER' in response:
            print("✅ Success: Server is advertising OAUTHBEARER!")
            
            # 3. Send SASLInitialResponse (KAG-8343 implementation)
            # This follows RFC 7628
            mechanism = b"OAUTHBEARER\0"
            sasl_payload = f"n,,auth=Bearer {TOKEN}\x01\x01".encode()
            
            # p + msg_len + mech_name + payload_len + payload
            p_msg = mechanism + struct.pack('!I', len(sasl_payload)) + sasl_payload
            final_packet = b'p' + struct.pack('!I', len(p_msg) + 4) + p_msg
            
            print(f"Sending RFC 7628 payload: {sasl_payload}")
            s.sendall(final_packet)
            
            # 4. Final Result
            result = s.recv(1024)
            print(f"Final Server Response: {result}")
        else:
            print("❌ Failure: OAUTHBEARER not in SASL list. Check pg_hba.conf.")
    
    s.close()

test_oauth_handshake()
